EDU-681 The "Update" feature (#2034)
temporalio/documentation: docs-src/go/how-to-handle-an-update-in-a-workflow-in-go.md

---
id: how-to-handle-an-update-in-go
title: How to handle an Update in Go
sidebar_label: Handle Update
description: Use the `SetUpateHandler` API from the `go.temporal.io/sdk/workflow` package to register an Update Handler for a given name.
tags:
  - go
  - how-to
---

> Use the `SetUpdateHandler` API from the `go.temporal.io/sdk/workflow` package to register an Update handler for a given name.

Reviewer: ```suggestion
Use the `SetUpdateHandler` API from the `go.temporal.io/sdk/workflow` package to register an Update handler for a given name.
```
Link to https://pkg.go.dev/go.temporal.io/sdk/workflow#SetUpdateHandler?

```go
func YourWorkflowDefinition(ctx workflow.Context, param YourWorkflowParam) error {
  counter := 0
  workflow.SetUpdateHandler(ctx, "fetchAdd", func(i int) (int, error) {
    tmp := counter
    counter += i
    return tmp, nil
  })
  // ...
}
```

> In the example above, the Workflow code uses `workflow.SetUpdateHandler` to register a function to handler workflow updates.
> The function can take multiple serializable input parameters though we recommend that only a single parameter is used to allow for fields to be added in future versions while retaining backward compatibility.

Reviewer: ```suggestion
In the preceding example, the Workflow code uses `workflow.SetUpdateHandler` to register a function to handle Workflow Updates.
The function can take multiple serializable input parameters, although we recommend that you use only a single parameter to allow for fields to be added in future versions while retaining backward compatibility.
```

> The function can optionally take a `workflow.Context` parameter in the first position.

Reviewer 2: @mmcshane - to save me from looking, we execute updates in a different coroutine right? To avoid issues/deadlocks in Go SDK, users _have_ to use the context for the coroutine they are in if they do anything async. So we should make it very clear in docs that if users are doing anything with a context, they _must_ take it in the function.

Sorry I forgot this point during original design. I am tempted to say context should be required now, but I can understand if too late. But if they mess up and use the outer context inside the update handler, their workflow can error in strange ways.

Reviewer 3: > we execute updates in a different coroutine

correct.

> users have to use the context for the coroutine they are in if they do anything async

also correct

Too late to change if we want to maintain compile compat with handlers written for older sdk-go versions. I can be convinced but I don't see this as a big enough deal to break that compatibility.

Reviewer 2: Not too concerned with compatibility at this stage.

I think a simple example of executing an activity outside of the update handler and inside the update handler both on the same outer `ctx` can demonstrate probably this issue. That is so easy to do and is an easy mistake to make. I am now starting to think we may need to require context as the first param for every update handler and I regret not catching it earlier. May want to consult with @mfateev here though.

The function returns either a serializable value and an error or just an error.

> Unlike with query handlers, it is safe for Update Handlers to observe and mutate workflow state.

Reviewer: ```suggestion
Unlike Query handlers, Update handlers can safely observe and mutate Workflow state.
```
